onnxruntime/core/codegen/target/tvm_scheduler.h

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/common/common.h"
+#include "core/codegen/common/creator.h"
+#include "core/codegen/common/registry.h"
+#include "core/codegen/target/tvm_context.h"
+#include "core/graph/graph.h"
+#include <tvm/tvm.h>
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+// These are current generic TVMOpRule we used.
+enum class TVMOpRuleType : int {
+  Extern = 0,
+  ComputeReduce = 1,
+  ComputeRegular = 2,
+  AlwaysRoot = 3,  // for debug
+  NoRule,
+};
+
+const std::string& GetTVMOpRule(const tvm::Tensor& tensor);
+const std::string& GetTVMOpRule(TVMOpRuleType rule);
+
+// These are current generic ScheduleType in tvm_codegen
+enum class ScheduleType : int {
+  ScheduleNone = 0,
+  ScheduleInline = 1,
+  ScheduleAt = 2,
+  ScheduleRoot = 3,
+  ScheduleClosure = 4,
+};
+
+// Data struct to bundle tvm::Schedule and scheduled tensor
+struct ScheduleContext {
+  ScheduleContext(const tvm::Array<tvm::Operation>& ops) {
+    schedule = tvm::create_schedule(ops);
+  }
+  tvm::Schedule schedule;
+  std::map<const tvm::Node*, ScheduleType> scheduled_tensors;
+};
+
+// Scheduler inserts a tvm::Schedule content to a tvm::Tensor
+using Scheduler = codegen::CreatorBase<
+    const tvm::Tensor&,
+    const Node*,
+    tvm_codegen::CodeGenContext&,
+    ScheduleContext&,
+    bool>;
+
+// TVMScheduleDispatcher is the base dispatcher for TVM Schedule Builder
+// It checks whether a pair of {tvm::Tensor, Ort Node} satisfying a criteria (in Find)
+// and dispatches a corresponding Scheduler.
+class TVMScheduleDispatcher : public codegen::DispatcherBase<Scheduler*> {
+ public:
+  TVMScheduleDispatcher(const std::string& name)
+      : DispatcherBase(name) {}
+
+  ~TVMScheduleDispatcher() = default;
+
+  virtual Scheduler* Find(const tvm::Tensor&,
+                          const Node*,
+                          tvm_codegen::CodeGenContext&) = 0;
```

Comment:
so the key to find a schedule pass is ""Tensor + Node + CodeGenContext"? it looks like only suitable for per-op schedule.

Reply:
It is not really per-op schedule. Context can hold a graph and corresponding analysis.
Tensor can be used as Key to get previous analyzed result and applied to schedule.

Reply:
As Li-wen mentioned, context can have more complicated schedule if needed. Please check Nuphar implementations as example.

---
In reply to: [294452737](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294452737) [](ancestors = 294452737)

Reply:
like mentioned in another comments, if that's the case, why you list tvm::Tensor here, context could cover everything. It makes the interface more confusing.

---
In reply to: [297449610](https://github.com/microsoft/onnxruntime/pull/888#discussion_r297449610) [](ancestors = 297449610,294452737)